### ISSUE TITLE
Saving bin image corrupts sectors

### ISSUE THREAD

**Issue by DLPB2** [myst6re/makoureactor#51]
This leads to the game being unusable in the emulator Mednafen

**Comment by myst6re** [comment]
True, everything we need to know is in [ECMA-119](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-119.pdf) standard. But that's a bit much to learn.

Note that some sectors at the begining of PS1 cds are not verified with ECMA-119 checksums, I think it's the anti-copy mechanism.

**Comment by DLPB2** [comment]
DLPB:  Replace portfile.cmake in vcpkg\ports\ff7tk

and add psx-cdrom-edc-ecc.h to vcpkg\ports\ff7tk


[ff7tk.zip](https://github.com/user-attachments/files/30881329/ff7tk.zip)

ChatGPT: That author response is only partly on target. I’m going to trace the image-writing code and compare it against the actual PlayStation CD sector formats, because a PS1 image can be perfectly valid as an ISO-9660 filesystem yet still be unusable if the 2352-byte raw sector structure, EDC/ECC, subheaders, or pregap/protection sectors are rewritten incorrectly.


One safety improvement while correcting this: relocated PS1 Mode 2 sectors now retain their original EDC/ECC bytes, because their MSF address is outside the Form 1 EDC and is zeroed for ECC calculation. That means deliberate/non-standard integrity bytes aren't unnecessarily “corrected” merely because a raw sector moved. Only sectors whose payload Makou actually changes get regenerated.


I tracked down the PSX disc-image corruption when saving from Makou Reactor. The fix is now tested: an FF7 image saved with the changes boots correctly in an emulator.

## Cause

The problem is in ff7tk's raw CD sector writing rather than the ISO-9660/ECMA-119 filesystem handling itself.

FF7 PSX filesystem data is stored in raw 2352-byte CD-ROM XA Mode 2 Form 1 sectors.

A Mode 2 Form 1 sector is effectively:

```text
24 bytes   sync/header/XA subheader
2048 bytes user data
4 bytes    EDC
172 bytes  ECC P
104 bytes  ECC Q
--------------------------------
2352 bytes total
```

The existing ff7tk code builds the first 24 bytes and the 2048-byte payload correctly, but `buildFooter()` does this:

```cpp
//TODO (if possible): Checksum EDC/ECC
// (Error Detection Code & Error Correction Code)

return QByteArray(SECTOR_SIZE_FOOTER, '\x00');
```

`SECTOR_SIZE_FOOTER` is 280 bytes.

Therefore every sector rebuilt by `writeSector()` was being written as:

```text
24-byte header
2048-byte data
280 zero bytes
```

instead of having valid EDC/P/Q ECC data.

This produces a filesystem which may still look structurally valid at the ISO-9660 level, but the raw PSX sectors themselves are invalid. Some emulators/drives tolerate this more than others.

There were two additional related problems.

### `writeIso()`

`writeIso()` modifies the 2048-byte payload inside an existing raw sector but did not recalculate the EDC/ECC afterwards.

This left sectors containing:

```text
new payload
old EDC/ECC
```

which is also invalid.

### `copySectors(..., repair=true)`

The existing repair path discarded most of the original raw sector and reconstructed it from the 2048-byte user-data area.

This unnecessarily destroys the original raw-sector integrity information and could also destroy intentional/non-standard sector data.

This matters particularly on PS1 discs because sectors which Makou does not modify should not be globally "corrected". Any unusual/protection/system sectors should remain byte-for-byte identical to the source disc.

## Fix

I added a CD-ROM EDC/ECC implementation and changed the writer so only sectors whose payload is actually modified have their integrity fields regenerated.

### Mode 2 Form 1

For XA Mode 2 Form 1 sectors the code now:

* calculates the 4-byte EDC;
* generates ECC P;
* generates ECC Q;
* correctly treats bytes 12-15 as zero while calculating XA Form 1 P/Q parity;
* restores the actual MSF/mode bytes afterwards.

### Mode 2 Form 2

Mode 2 Form 2 sectors are handled separately:

* the Form 2 EDC is regenerated;
* no P/Q ECC is generated because Form 2 does not contain it.

### Mode 1

The helper also handles Mode 1 sectors correctly if encountered.

## Existing-sector writes

After `writeIso()` changes data inside one or more raw sectors, every affected sector now has its EDC/ECC regenerated before the operation completes.

This also fixes ISO filesystem metadata modifications such as directory-record updates.

The volume-space-size writes in the Primary Volume Descriptor were previously made directly with:

```cpp
destinationIO->write(...)
```

They are now routed through:

```cpp
destinationIO->writeIso(...)
```

so those sectors receive the same EDC/ECC update.

## Sector relocation

`copySectors(..., repair=true)` now keeps the complete original 2352-byte raw sector instead of extracting only the 2048-byte payload.

When relocating an existing sector:

* the raw sector is preserved;
* the destination MSF address is updated where appropriate;
* Mode 2 XA EDC/ECC bytes are otherwise preserved exactly.

For XA Mode 2 Form 1, the EDC starts at the XA subheader rather than including the MSF header, and the address bytes are treated as zero for P/Q ECC calculation. Therefore merely changing the physical MSF position does not require regenerating the existing XA Form 1 integrity data.

This also means any intentional/non-standard EDC/ECC values remain untouched when a sector is merely relocated.

Mode 1 is different because its integrity calculation includes the address, so Mode 1 sectors are regenerated after relocation.

## Important PS1 behaviour

There is deliberately no whole-disc EDC/ECC repair pass.

Untouched 2352-byte sectors are copied verbatim.

So:

```text
Untouched sector
    -> copy all 2352 bytes unchanged
```

whereas:

```text
Sector whose payload Makou changes
    -> modify payload
    -> regenerate correct EDC/ECC
```

This preserves the PS1 system area and any intentionally unusual sectors while still producing valid raw sectors for data Makou actually changes.

## Files changed in my Makou Reactor fork

Because ff7tk is consumed through the local vcpkg overlay, the final working version changes/adds only these two project files:

```text
MODIFIED:
.vcpkg/ports/ff7tk/portfile.cmake

ADDED:
.vcpkg/ports/ff7tk/psx-cdrom-edc-ecc.h
```

### `.vcpkg/ports/ff7tk/psx-cdrom-edc-ecc.h`

New helper containing:

* CD-ROM EDC lookup/calculation;
* ECC forward/backward tables;
* P/Q ECC generation;
* Mode 1 integrity generation;
* XA Mode 2 Form 1 EDC/P/Q generation;
* XA Mode 2 Form 2 EDC generation;
* existing-sector repair;
* safe raw-sector relocation.

### `.vcpkg/ports/ff7tk/portfile.cmake`

The overlay still downloads pristine ff7tk v1.3.1.

At build time it copies the EDC/ECC helper into:

```text
src/formats/psx-cdrom-edc-ecc.h
```

and applies the required changes to `src/formats/IsoArchive.cpp`.

The affected ff7tk functions are principally:

```text
IsoArchiveIO::writeIso()
IsoArchiveIO::writeSector()
IsoArchive::copySectors()
```

plus the Primary Volume Descriptor volume-size writes.

For an upstream ff7tk fix, the vcpkg transformation itself obviously isn't required; the equivalent changes can be incorporated directly into `IsoArchive.cpp/.h` and the EDC/ECC helper can live normally within ff7tk.

## Result

Before the change:

```text
Makou save
 -> modified Mode 2 Form 1 sector
 -> 280-byte footer written as zeros / stale integrity data
 -> resulting PSX image can fail to boot
```

After the change:

```text
Makou save
 -> modified sector
 -> correct EDC
 -> correct ECC P
 -> correct ECC Q
 -> untouched sectors preserved verbatim
 -> saved FF7 image boots correctly
```

So the corruption was not caused by ISO-9660/ECMA-119 directory layout itself; it was caused by ff7tk not generating the raw CD-ROM XA sector integrity fields after changing sector data.

**Comment by DLPB2** [comment]
@myst6re   this has been verified to work on my build. I loaded Japanese iso that had been saved with this code and it the iso played correctly without crash in Mednafen.

**Comment by DLPB2** [comment]
Added to https://github.com/sithlord48/ff7tk/pull/215
